File: gcp/workers/recoverer/recoverer.py
#!/usr/bin/env python3
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""OSV failed task recoverer."""

import base64
import datetime
import logging
import os
import sys
import time

from google.cloud import ndb
from google.cloud import pubsub_v1

import osv
import osv.models
import osv.sources
from osv.logs import setup_gcp_logging

_FAILED_TASKS_SUBSCRIPTION = 'recovery'
_TASKS_TOPIC = 'tasks'

_ndb_client = None


def ndb_client():
  """Get the ndb client.
  Lazily initialized to allow testing with datastore emulator."""
  global _ndb_client
  if _ndb_client is None:
    _ndb_client = ndb.Client()
  return _ndb_client


def handle_gcs_retry(message: pubsub_v1.types.PubsubMessage) -> bool:
  """Handle a failed GCS write."""
  try:
    vuln = osv.vulnerability_pb2.Vulnerability.FromString(message.data)
  except Exception:
    logging.error(
        'gcs_retry: failed to decode protobuf. Ignoring message.',
        # chuck the data into the GCP log fields in case it's useful.
        extra={
            'json_fields': {
                'data': base64.encodebytes(message.data).decode()
            }
        })
    return True
  logging.info('gcs_retry: vulnerability: %s', vuln.id)
  modified = vuln.modified.ToDatetime(datetime.UTC)
  bucket = osv.gcs.get_osv_bucket()
  path = os.path.join(osv.gcs.VULN_PB_PATH, vuln.id + '.pb')
  pb_blob = bucket.get_blob(path)
  # Check that the record hasn't been written/updated in the meantime.
  if pb_blob and pb_blob.custom_time and pb_blob.custom_time >= modified:
    logging.warning(
        'gcs_retry: %s was modified before message was processed: '
        'message: %s, blob: %s', vuln.id, modified, pb_blob.custom_time)
    # TODO(michaelkedar): trigger a reimport of the record.
    return True

  pb_blob = bucket.blob(path)
  pb_blob.custom_time = modified
  try:
    pb_blob.upload_from_string(
        message.data, content_type='application/octet-stream')
    return True
  except Exception:
    logging.exception('gcs_retry: failed to upload %s protobuf to GCS', vuln.id)
    return False


def handle_gcs_missing(message: pubsub_v1.types.PubsubMessage) -> bool:
  """Handle a failed GCS read."""
  vuln_id = message.attributes.get('id')
  logging.info('gcs_missing: vulnerability: %s', vuln_id)
  if not vuln_id:
    logging.error('gcs_missing: message missing id attribute: %s', message)
    return True

  with ndb_client().context():
    vuln = osv.Vulnerability.get_by_id(vuln_id)
    if not vuln:
      logging.error('gcs_missing: Vulnerability entity not found for %s',
                    vuln_id)
      return True

    try:
      source, path = osv.sources.parse_source_id(vuln.source_id)
    except ValueError:
      logging.error('gcs_missing: invalid source_id for %s: %s', vuln_id,
                    vuln.source_id)
      return True

    logging.info('gcs_missing: triggering re-import for %s (%s)', vuln_id,
                 vuln.source_id)
    publisher = pubsub_v1.PublisherClient()
    project = os.environ['GOOGLE_CLOUD_PROJECT']
    topic_path = publisher.topic_path(project, _TASKS_TOPIC)
    publisher.publish(
        topic_path,
        data=b'',
        type='update',
        source=source,
        path=path,
        original_sha256='',
        deleted='false',
        skip_hash_check='true',
        req_timestamp=str(int(time.time())))

    return True


def handle_gcs_gen_mismatch(message: pubsub_v1.types.PubsubMessage) -> bool:
  """Handle a generation mismatch when attempting update a part of a record.
  e.g. If a record was reimported while its aliases were being updated.
  """
  vuln_id = message.attributes.get('id')
  field = message.attributes.get('field')
  logging.info('gcs_gen_mismatch: vulnerability: %s, field: %s', vuln_id, field)
  if not vuln_id or not field:
    logging.error('gcs_gen_mismatch: message missing id or field attribute: %s',
                  message)
    return True

  with ndb_client().context():
    result = osv.gcs.get_by_id_with_generation(vuln_id)
    if result is None:
      logging.error('gcs_gen_mismatch: vulnerability not in GCS - %s', vuln_id)
      logging.info('trying with gcs_missing')
      return handle_gcs_missing(message)
    vuln_proto, generation = result

    def transaction():
      vuln: osv.Vulnerability = osv.Vulnerability.get_by_id(vuln_id)
      if vuln is None:
        logging.error('vulnerability not in Datastore - %s', vuln_id)
        # TODO(michaelkedar): What to do in this case?
        return
      modified = vuln.modified

      fields = field.split(',')
      for f in fields:
        if f == 'aliases':
          alias_group = osv.AliasGroup.query(
              osv.AliasGroup.bug_ids == vuln_id).get()
          if alias_group is None:
            aliases = []
            aliases_modified = datetime.datetime.now(datetime.UTC)
          else:
            aliases = sorted(set(alias_group.bug_ids) - {vuln_id})
            aliases_modified = alias_group.last_modified
          # Only update the modified time if it's actually being modified
          if vuln_proto.aliases != aliases:
            vuln_proto.aliases[:] = aliases
            if aliases_modified > modified:
              modified = aliases_modified
            else:
              modified = datetime.datetime.now(datetime.UTC)

        elif f == 'upstream':
          upstream_group = osv.UpstreamGroup.query(
              osv.UpstreamGroup.db_id == vuln_id).get()
          if upstream_group is None:
            upstream = []
            upstream_modified = datetime.datetime.now(datetime.UTC)
          else:
            upstream = upstream_group.upstream_ids
            upstream_modified = upstream_group.last_modified
          # Only update the modified time if it's actually being modified
          if vuln_proto.upstream != upstream:
            vuln_proto.upstream[:] = upstream
            if upstream_modified > modified:
              modified = upstream_modified
            else:
              modified = datetime.datetime.now(datetime.UTC)

      vuln_proto.modified.FromDatetime(modified)
      osv.ListedVulnerability.from_vulnerability(vuln_proto).put()
      vuln.modified = modified
      vuln.put()

    try:
      ndb.transaction(transaction)
    except Exception:
      logging.exception(
          'gcs_gen_mismatch: Datastore transaction failed for %s %s', vuln_id,
          field)
      return False
    try:
      osv.gcs.upload_vulnerability(vuln_proto, generation)
      return True
    except Exception:
      logging.exception('gcs_gen_mismatch: Writing to bucket failed for %s %s',
                        vuln_id, field)
      return False


def handle_generic(message: pubsub_v1.types.PubsubMessage) -> bool:
  """Generic message handler."""
  task_type = message.attributes.get('type', 'unknown')
  logging.error('`%s` task could not be processed: %s', task_type, message)
  # TODO(michaelkedar): We should store these somewhere.
  return True


HANDLERS = {
    'gcs_retry': handle_gcs_retry,
    'gcs_missing': handle_gcs_missing,
    'gcs_gen_mismatch': handle_gcs_gen_mismatch,
}


def handle_task(message: pubsub_v1.types.PubsubMessage) -> bool:
  """Handle a 'failed-tasks' message."""
  task_type = message.attributes.get('type')
  handler = HANDLERS.get(task_type, handle_generic)
  return handler(message)


def main():
  project = osv.utils.get_google_cloud_project()
  if not project:
    logging.error('GOOGLE_CLOUD_PROJECT not set')
    sys.exit(1)

  with pubsub_v1.SubscriberClient() as subscriber:
    subscription = subscriber.subscription_path(project,
                                                _FAILED_TASKS_SUBSCRIPTION)

    while True:
      response = subscriber.pull(subscription=subscription, max_messages=1)
      if not response.received_messages:
        continue

      message = response.received_messages[0].message
      ack_id = response.received_messages[0].ack_id
      # Try handle the task
      # If successful (returned True), acknowledge it.
      # Otherwise, nack the task to trigger it to be redelivered.
      if handle_task(message):
        subscriber.acknowledge(subscription=subscription, ack_ids=[ack_id])
      else:
        subscriber.modify_ack_deadline(
            subscription=subscription, ack_ids=[ack_id], ack_deadline_seconds=0)


if __name__ == '__main__':
  setup_gcp_logging('recoverer')
  main()

File: go/internal/recoverer/recoverer.go
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Package recoverer implements the OSV failed task recoverer.
package recoverer

import (
	"bytes"
	"context"
	"encoding/base64"
	"errors"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"net/url"
	"slices"
	"strconv"
	"strings"
	"time"

	"cloud.google.com/go/datastore"
	"cloud.google.com/go/pubsub/v2"
	osvdatastore "github.com/google/osv.dev/go/internal/database/datastore"
	"github.com/google/osv.dev/go/internal/gitter"
	gitterpb "github.com/google/osv.dev/go/internal/gitter/pb/repository"
	"github.com/google/osv.dev/go/internal/models"
	"github.com/google/osv.dev/go/logger"
	"github.com/google/osv.dev/go/osv/clients"
	"github.com/ossf/osv-schema/bindings/go/osvschema"
	"github.com/tidwall/gjson"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/trace"
	"google.golang.org/protobuf/encoding/protojson"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/timestamppb"
	"k8s.io/apimachinery/pkg/util/yaml"
)

type Stores struct {
	DatastoreClient *datastore.Client
	SourceRepo      models.SourceRepositoryStore
	Relations       models.RelationsStore
	GCS             clients.CloudStorage
	GCSProvider     clients.CloudStorageProvider
	Publisher       clients.Publisher
}

type Config struct {
	Stores           Stores
	GitterClient     gitter.Client
	HTTPClient       *http.Client
	DefaultTaskPool  string
	ReimportTaskPool string
}

type Recoverer struct {
	stores           Stores
	gitterClient     gitter.Client
	httpClient       *http.Client
	defaultTaskPool  string
	reimportTaskPool string
}

func New(cfg Config) *Recoverer {
	defaultPool := cfg.DefaultTaskPool
	if defaultPool == "" {
		defaultPool = "default"
	}
	reimportPool := cfg.ReimportTaskPool
	if reimportPool == "" {
		reimportPool = "reimport"
	}
	httpClient := cfg.HTTPClient
	if httpClient == nil {
		httpClient = http.DefaultClient
	}

	return &Recoverer{
		stores:           cfg.Stores,
		gitterClient:     cfg.GitterClient,
		httpClient:       httpClient,
		defaultTaskPool:  defaultPool,
		reimportTaskPool: reimportPool,
	}
}

// Run starts the subscriber loop listening for recovery tasks.
func (r *Recoverer) Run(ctx context.Context, sub *pubsub.Subscriber) error {
	return sub.Receive(ctx, r.HandleMessage)
}

// HandleMessage handles a single Pub/Sub message from the failed tasks subscription.
func (r *Recoverer) HandleMessage(ctx context.Context, m *pubsub.Message) {
	taskType := m.Attributes["type"]
	logInfo := []any{
		slog.String("type", taskType),
		slog.String("id", m.Attributes["id"]),
	}

	taskCtx := otel.GetTextMapPropagator().Extract(ctx, propagation.MapCarrier(m.Attributes))
	taskCtx, span := otel.Tracer("recoverer").Start(taskCtx, "process_task",
		trace.WithAttributes(attribute.String("task_type", taskType)))
	defer span.End()

	var err error
	switch taskType {
	case "gcs_retry":
		err = r.HandleGCSRetry(taskCtx, m)
	case "gcs_missing":
		err = r.HandleGCSMissing(taskCtx, m)
	case "gcs_gen_mismatch":
		err = r.HandleGCSGenMismatch(taskCtx, m)
	default:
		err = r.HandleGeneric(taskCtx, m)
	}

	if err != nil {
		logger.ErrorContext(taskCtx, "Failed to process recovery task", append(logInfo, slog.Any("error", err))...)
		m.Nack()
	} else {
		m.Ack()
	}
}

// HandleGCSRetry handles a failed GCS write.
func (r *Recoverer) HandleGCSRetry(ctx context.Context, m *pubsub.Message) error {
	var vuln osvschema.Vulnerability
	if err := proto.Unmarshal(m.Data, &vuln); err != nil {
		logger.ErrorContext(ctx, "gcs_retry: failed to decode protobuf. Ignoring message.",
			slog.Any("error", err),
			slog.String("data_base64", base64.StdEncoding.EncodeToString(m.Data)))
		// Non-retryable decoding error, return nil to ack
		return nil
	}

	vulnID := vuln.GetId()
	logger.InfoContext(ctx, "gcs_retry: vulnerability", slog.String("id", vulnID))
	modified := vuln.GetModified().AsTime()
	path := fmt.Sprintf("all/pb/%s.pb", vulnID)

	attrs, err := r.stores.GCS.ReadObjectAttrs(ctx, path)
	if err == nil {
		if !attrs.CustomTime.IsZero() && (attrs.CustomTime.Equal(modified) || attrs.CustomTime.After(modified)) {
			logger.WarnContext(ctx, "gcs_retry: was modified before message was processed",
				slog.String("id", vulnID),
				slog.Time("message_modified", modified),
				slog.Time("blob_modified", attrs.CustomTime))

			return nil
		}
	} else if !errors.Is(err, clients.ErrNotFound) {
		logger.ErrorContext(ctx, "gcs_retry: failed to get object attrs from GCS", slog.String("id", vulnID), slog.Any("error", err))

		return err
	}

	opts := &clients.WriteOptions{
		CustomTime:  &modified,
		ContentType: "application/octet-stream",
	}
	if err := r.stores.GCS.WriteObject(ctx, path, m.Data, opts); err != nil {
		logger.ErrorContext(ctx, "gcs_retry: failed to upload protobuf to GCS", slog.String("id", vulnID), slog.Any("error", err))

		return err
	}

	return nil
}

// HandleGCSMissing handles a failed GCS read by triggering a reimport.
func (r *Recoverer) HandleGCSMissing(ctx context.Context, m *pubsub.Message) error {
	vulnID := m.Attributes["id"]
	logger.InfoContext(ctx, "gcs_missing: vulnerability", slog.String("id", vulnID))
	if vulnID == "" {
		logger.ErrorContext(ctx, "gcs_missing: message missing id attribute")

		return nil
	}

	var vuln osvdatastore.Vulnerability
	key := datastore.NameKey("Vulnerability", vulnID, nil)
	if err := r.stores.DatastoreClient.Get(ctx, key, &vuln); err != nil {
		if errors.Is(err, datastore.ErrNoSuchEntity) {
			logger.ErrorContext(ctx, "gcs_missing: Vulnerability entity not found", slog.String("id", vulnID))

			return nil
		}
		logger.ErrorContext(ctx, "gcs_missing: failed to fetch Vulnerability entity", slog.String("id", vulnID), slog.Any("error", err))

		return err
	}

	source, path, ok := strings.Cut(vuln.SourceID, ":")
	if !ok || source == "" || path == "" {
		logger.ErrorContext(ctx, "gcs_missing: invalid source_id", slog.String("id", vulnID), slog.String("source_id", vuln.SourceID))

		return nil
	}

	logger.InfoContext(ctx, "gcs_missing: triggering re-import", slog.String("id", vulnID), slog.String("source_id", vuln.SourceID))

	data, err := r.DownloadVulnData(ctx, source, path)
	if err != nil {
		logger.ErrorContext(ctx, "gcs_missing: failed to download vuln data", slog.String("id", vulnID), slog.String("source_id", vuln.SourceID), slog.Any("error", err))

		return err
	}

	msg := &pubsub.Message{
		Data: data,
		Attributes: map[string]string{
			"type":             "update",
			"source":           source,
			"path":             path,
			"original_sha256":  "",
			"deleted":          "false",
			"skip_hash_check":  "true",
			"req_timestamp":    strconv.FormatInt(time.Now().Unix(), 10),
			"content_encoding": "",
			"work_pool":        r.reimportTaskPool,
		},
	}
	otel.GetTextMapPropagator().Inject(ctx, propagation.MapCarrier(msg.Attributes))

	res := r.stores.Publisher.Publish(ctx, msg)
	if _, err := res.Get(ctx); err != nil {
		logger.ErrorContext(ctx, "gcs_missing: failed to publish update message", slog.String("id", vulnID), slog.Any("error", err))

		return err
	}

	return nil
}

// DownloadVulnData downloads and parses vulnerability data from the source repository.
func (r *Recoverer) DownloadVulnData(ctx context.Context, source, vulnPath string) ([]byte, error) {
	sourceRepo, err := r.stores.SourceRepo.Get(ctx, source)
	if err != nil {
		return nil, fmt.Errorf("failed to get source repository %s: %w", source, err)
	}

	var rawData []byte
	switch sourceRepo.Type {
	case models.SourceRepositoryTypeREST:
		if sourceRepo.Link == "" {
			return nil, fmt.Errorf("REST source repository %s is missing Link", source)
		}
		reqURL, err := url.JoinPath(sourceRepo.Link, vulnPath)
		if err != nil {
			return nil, fmt.Errorf("failed to construct REST URL for %s / %s: %w", sourceRepo.Link, vulnPath, err)
		}
		req, err := http.NewRequestWithContext(ctx, http.MethodGet, reqURL, nil)
		if err != nil {
			return nil, fmt.Errorf("failed to create HTTP request: %w", err)
		}
		resp, err := r.httpClient.Do(req)
		if err != nil {
			return nil, fmt.Errorf("HTTP request failed for %s: %w", reqURL, err)
		}
		defer resp.Body.Close()
		if resp.StatusCode != http.StatusOK {
			return nil, fmt.Errorf("HTTP request failed with status %d for %s", resp.StatusCode, reqURL)
		}
		rawData, err = io.ReadAll(resp.Body)
		if err != nil {
			return nil, fmt.Errorf("failed to read response body for %s: %w", reqURL, err)
		}

	case models.SourceRepositoryTypeBucket:
		if sourceRepo.Bucket == nil || sourceRepo.Bucket.Name == "" {
			return nil, fmt.Errorf("bucket source repository %s is missing bucket name", source)
		}
		if r.stores.GCSProvider == nil {
			return nil, errors.New("GCSProvider is not configured")
		}
		bucketStorage := r.stores.GCSProvider.Bucket(sourceRepo.Bucket.Name)
		rawData, err = bucketStorage.ReadObject(ctx, vulnPath)
		if err != nil {
			return nil, fmt.Errorf("failed to read object %s from bucket %s: %w", vulnPath, sourceRepo.Bucket.Name, err)
		}

	case models.SourceRepositoryTypeGit:
		if sourceRepo.Git == nil || sourceRepo.Git.URL == "" {
			return nil, fmt.Errorf("git source repository %s is missing git url", source)
		}
		if r.gitterClient == nil {
			return nil, errors.New("gitter client is not configured")
		}
		ref := sourceRepo.Git.LastSyncedCommit
		if ref == "" {
			ref = sourceRepo.Git.Branch
		}
		if ref == "" {
			ref = "HEAD"
		}
		req := &gitterpb.FileContentRequest{
			Url:    sourceRepo.Git.URL,
			Commit: ref,
			Path:   vulnPath,
		}
		resp, err := r.gitterClient.GetFileContent(ctx, req)
		if err != nil {
			return nil, fmt.Errorf("gitter GetFileContent failed for %s (commit %s, path %s): %w", sourceRepo.Git.URL, ref, vulnPath, err)
		}
		rawData = resp.GetContent()

	default:
		return nil, fmt.Errorf("unhandled source repository type: %v", sourceRepo.Type)
	}

	vulnProto, err := parseVulnerability(rawData, sourceRepo.Extension, vulnPath, sourceRepo.KeyPath)
	if err != nil {
		return nil, fmt.Errorf("failed to parse vulnerability data for %s:%s: %w", source, vulnPath, err)
	}

	return proto.Marshal(vulnProto)
}

func parseVulnerability(raw []byte, extension, vulnPath, keyPath string) (*osvschema.Vulnerability, error) {
	data := bytes.ToValidUTF8(raw, []byte("\uFFFD"))
	ext := strings.ToLower(extension)
	if ext == ".yaml" || ext == ".yml" || strings.HasSuffix(strings.ToLower(vulnPath), ".yaml") || strings.HasSuffix(strings.ToLower(vulnPath), ".yml") {
		jsonBytes, err := yaml.ToJSON(data)
		if err != nil {
			return nil, fmt.Errorf("failed to convert YAML to JSON: %w", err)
		}
		data = jsonBytes
	}

	if keyPath != "" {
		res := gjson.GetBytes(data, keyPath)
		if !res.Exists() {
			return nil, fmt.Errorf("key path %q not found in data", keyPath)
		}
		data = []byte(res.Raw)
	}

	var vulnProto osvschema.Vulnerability
	unmarshalOptions := protojson.UnmarshalOptions{DiscardUnknown: true}
	if err := unmarshalOptions.Unmarshal(data, &vulnProto); err != nil {
		return nil, fmt.Errorf("failed to unmarshal OSV proto: %w", err)
	}

	return &vulnProto, nil
}

// HandleGCSGenMismatch handles a generation mismatch when attempting to update a part of a record.
func (r *Recoverer) HandleGCSGenMismatch(ctx context.Context, m *pubsub.Message) error {
	vulnID := m.Attributes["id"]
	fieldStr := m.Attributes["field"]
	logger.InfoContext(ctx, "gcs_gen_mismatch: vulnerability", slog.String("id", vulnID), slog.String("field", fieldStr))
	if vulnID == "" || fieldStr == "" {
		logger.ErrorContext(ctx, "gcs_gen_mismatch: message missing id or field attribute")

		return nil
	}

	path := fmt.Sprintf("all/pb/%s.pb", vulnID)
	attrs, err := r.stores.GCS.ReadObjectAttrs(ctx, path)
	if errors.Is(err, clients.ErrNotFound) {
		logger.ErrorContext(ctx, "gcs_gen_mismatch: vulnerability not in GCS", slog.String("id", vulnID))
		logger.InfoContext(ctx, "trying with gcs_missing", slog.String("id", vulnID))

		return r.HandleGCSMissing(ctx, m)
	}
	if err != nil {
		logger.ErrorContext(ctx, "gcs_gen_mismatch: failed to read object attrs from GCS", slog.String("id", vulnID), slog.Any("error", err))

		return err
	}
	generation := attrs.Generation

	data, err := r.stores.GCS.ReadObject(ctx, path)
	if err != nil {
		logger.ErrorContext(ctx, "gcs_gen_mismatch: failed to read object data from GCS", slog.String("id", vulnID), slog.Any("error", err))

		return err
	}

	var baseProto osvschema.Vulnerability
	if err := proto.Unmarshal(data, &baseProto); err != nil {
		logger.ErrorContext(ctx, "gcs_gen_mismatch: failed to unmarshal proto from GCS", slog.String("id", vulnID), slog.Any("error", err))

		return nil
	}

	var modified time.Time
	var finalProto *osvschema.Vulnerability
	vulnKey := datastore.NameKey("Vulnerability", vulnID, nil)
	dsFound := true

	_, txErr := r.stores.DatastoreClient.RunInTransaction(ctx, func(tx *datastore.Transaction) error {
		var dsVuln osvdatastore.Vulnerability
		if err := tx.Get(vulnKey, &dsVuln); err != nil {
			if errors.Is(err, datastore.ErrNoSuchEntity) {
				logger.ErrorContext(ctx, "vulnerability not in Datastore", slog.String("id", vulnID))
				dsFound = false

				return nil
			}

			return err
		}
		modified = dsVuln.Modified
		vulnProto := proto.Clone(&baseProto).(*osvschema.Vulnerability)

		fields := strings.Split(fieldStr, ",")
		for _, f := range fields {
			switch strings.TrimSpace(f) {
			case "aliases":
				aliasResult, err := r.stores.Relations.GetAliases(ctx, vulnID)
				var aliases []string
				var aliasesModified time.Time
				if errors.Is(err, models.ErrNotFound) {
					aliases = []string{}
					aliasesModified = time.Now().UTC()
				} else if err != nil {
					return fmt.Errorf("failed to get aliases for %s: %w", vulnID, err)
				} else {
					aliases = aliasResult.Aliases
					aliasesModified = aliasResult.Modified
				}
				if !slices.Equal(vulnProto.GetAliases(), aliases) {
					vulnProto.Aliases = aliases
					if aliasesModified.After(modified) {
						modified = aliasesModified
					} else {
						modified = time.Now().UTC()
					}
				}

			case "upstream":
				upstreamResult, err := r.stores.Relations.GetUpstream(ctx, vulnID)
				var upstream []string
				var upstreamModified time.Time
				if errors.Is(err, models.ErrNotFound) {
					upstream = []string{}
					upstreamModified = time.Now().UTC()
				} else if err != nil {
					return fmt.Errorf("failed to get upstream for %s: %w", vulnID, err)
				} else {
					upstream = upstreamResult.Upstream
					upstreamModified = upstreamResult.Modified
				}
				if !slices.Equal(vulnProto.GetUpstream(), upstream) {
					vulnProto.Upstream = upstream
					if upstreamModified.After(modified) {
						modified = upstreamModified
					} else {
						modified = time.Now().UTC()
					}
				}

			case "related":
				relatedResult, err := r.stores.Relations.GetRelated(ctx, vulnID)
				var related []string
				var relatedModified time.Time
				if errors.Is(err, models.ErrNotFound) {
					related = []string{}
					relatedModified = time.Now().UTC()
				} else if err != nil {
					return fmt.Errorf("failed to get related for %s: %w", vulnID, err)
				} else {
					related = relatedResult.Related
					relatedModified = relatedResult.Modified
				}
				if !slices.Equal(vulnProto.GetRelated(), related) {
					vulnProto.Related = related
					if relatedModified.After(modified) {
						modified = relatedModified
					} else {
						modified = time.Now().UTC()
					}
				}
			}
		}

		vulnProto.Modified = timestamppb.New(modified)
		dsVuln.Modified = modified
		listedVuln := osvdatastore.NewListedVulnerabilityFromProto(vulnProto)
		listedKey := datastore.NameKey("ListedVulnerability", vulnID, nil)

		if _, err := tx.Put(vulnKey, &dsVuln); err != nil {
			return err
		}
		if _, err := tx.Put(listedKey, listedVuln); err != nil {
			return err
		}

		finalProto = vulnProto

		return nil
	})

	if txErr != nil {
		logger.ErrorContext(ctx, "gcs_gen_mismatch: Datastore transaction failed",
			slog.String("id", vulnID), slog.String("field", fieldStr), slog.Any("error", txErr))

		return txErr
	}

	if !dsFound {
		return nil
	}

	newData, err := proto.Marshal(finalProto)
	if err != nil {
		logger.ErrorContext(ctx, "gcs_gen_mismatch: failed to marshal proto", slog.String("id", vulnID), slog.Any("error", err))

		return err
	}
	opts := &clients.WriteOptions{
		IfGenerationMatches: &generation,
		CustomTime:          &modified,
		ContentType:         "application/octet-stream",
	}
	if err := r.stores.GCS.WriteObject(ctx, path, newData, opts); err != nil {
		logger.ErrorContext(ctx, "gcs_gen_mismatch: Writing to bucket failed",
			slog.String("id", vulnID), slog.String("field", fieldStr), slog.Any("error", err))

		return err
	}

	return nil
}

// HandleGeneric handles unhandled task types by logging and acknowledging.
func (r *Recoverer) HandleGeneric(ctx context.Context, m *pubsub.Message) error {
	taskType := m.Attributes["type"]
	if taskType == "" {
		taskType = "unknown"
	}
	logger.ErrorContext(ctx, fmt.Sprintf("`%s` task could not be processed", taskType),
		slog.String("type", taskType),
		slog.Any("attributes", m.Attributes))

	return nil
}
